synapse/storage/databases/main/receipts.py

```diff
@@ -133,59 +133,95 @@ async def get_receipts_for_room(
+                self.database_engine, "rl.receipt_type", receipt_types
+            )
+            sql = """
+                SELECT t.room_id, t.event_id FROM (
+                    SELECT rl.room_id, rl.event_id, row_number() over (partition by rl.room_id order by e.stream_ordering desc) rw
+                    FROM receipts_linearized AS rl
+                    INNER JOIN events AS e USING (room_id, event_id)
+                    WHERE rl.user_id = ?
+                    AND %s
+                ) t
+                WHERE t.rw = 1;
+
+            """ % (
+                clause,
+            )
 
-        return {row["room_id"]: row["event_id"] for row in rows}
+            txn.execute(sql, [user_id] + list(args))
+            return cast(List[Tuple[str, str]], txn.fetchall())
+
+        rows = await self.db_pool.runInteraction("get_latest_receipts_for_user", f)
+        return {row[0]: row[1] for row in rows}
 
     async def get_receipts_for_user_with_orderings(
-        self, user_id: str, receipt_type: str
+        self, user_id: str, receipt_types: List[str]
     ) -> JsonDict:
-        def f(txn: LoggingTransaction) -> List[Tuple[str, str, int, int]]:
-            sql = (
-                "SELECT rl.room_id, rl.event_id,"
-                " e.topological_ordering, e.stream_ordering"
-                " FROM receipts_linearized AS rl"
-                " INNER JOIN events AS e USING (room_id, event_id)"
-                " WHERE rl.room_id = e.room_id"
-                " AND rl.event_id = e.event_id"
-                " AND user_id = ?"
+        def f(txn: LoggingTransaction) -> List[Tuple[str, str, str, int, int]]:
+            clause, args = make_in_list_sql_clause(
+                self.database_engine, "rl.receipt_type", receipt_types
             )
-            txn.execute(sql, (user_id,))
-            return cast(List[Tuple[str, str, int, int]], txn.fetchall())
+
+            sql = """
+                SELECT rl.room_id, rl.event_id, rl.receipt_type,
+                e.topological_ordering, e.stream_ordering
+                FROM receipts_linearized AS rl
+                INNER JOIN events AS e USING (room_id, event_id)
+                WHERE rl.room_id = e.room_id
+                AND rl.event_id = e.event_id
+                AND user_id = ?
+                AND %s
+            """ % (
+                clause,
+            )
+
+            txn.execute(sql, [user_id] + list(args))
+            return cast(List[Tuple[str, str, str, int, int]], txn.fetchall())
 
         rows = await self.db_pool.runInteraction(
             "get_receipts_for_user_with_orderings", f
         )
         return {
             row[0]: {
                 "event_id": row[1],
-                "topological_ordering": row[2],
-                "stream_ordering": row[3],
+                "receipt_type": row[2],
+                "topological_ordering": row[3],
+                "stream_ordering": row[4],
             }
             for row in rows
         }
```

Comment:
I don't think this is quite right, it will only return a single receipt per room, even though we're fetching (potentially) multiple receipts per room.